Ability to tweak socket/TLS parameters.

bryan-inomial opened the pull request:
# TLS fine-tuning

The fifth argument (`websocketClientConfig`) allows low-level socket/TLS parameters of the websocket connection to be fine-tuned. This argument if non-null should be a JS object; supported properties are documented [here](https://github.com/theturtle32/WebSocket-Node/blob/master/docs/WebSocketClient.md#client-config-options).

For example, supplying the below JS object for the `websocketClientConfig` argument:

  ```js
  {
    tlsOptions: {
      rejectUnauthorized: false
    }
  }
  ```

…will disable server TLS certificate validation (avoids having to set the `NODE_TLS_REJECT_UNAUTHORIZED` variable to `0` and the warning message from node.js it produces).

Alternatively supplying the following object for the `websocketClientConfig` argument:

  ```js
  {
    tlsOptions: {
      ca: /* Insert contents of X.509 custom root CA certificate in PEM format here as a string */
    }
  }
  ```

…will allow the server TLS certificate to be validated against that custom root CA certificate (even if it's not listed in the default trust store).

# Caveats

It was suggested by Mark that I could add a `setClientConfig()` method instead of adding a fifth argument to the exported `connect()` function. However I'm unable to do this as the `connect()` function itself tries to initiate the connection and return the `InomialClient` instance in the connected state, at which point it's too late to supply alternate TLS parameters.

A better suggestion would be to have the `InomialClient.js` module export the `InomialClient` class instead of the `connect()` function, and then have separate `connect()` and `setClientConfig()` methods; this will avoid having to add a fifth argument to the `connect()` function.

bryan-inomial committed:
Ability to tweak TLS params & graduated log levels

InomialClient constructor now accepts a fifth argument that
allows socket/TLS parameters to be fine tuned, and a sixth
argument that specifies the logging level (from none up to
fine wire-level logging).

Prefixed all logging messages produced by InomialClient
with [InomialClient], so it's clear where the messages
are coming from in the developer's application.

Tidied up README.md, Javascript snippets are now marked
to allow syntax highlighting in Github's Markdown renderer.

Also make InomialClient constructor throw an exception if the
hostname or stage are completely unavailable (caller has
passed in null and the respective environment variables aren't set).

bryan-inomial committed:
Revise FINE-level request query logging algorithm

Came up with a better approach: if a large query document and operation name is
given, then use a quick-and-dirty regex search to locate the
query/mutation/subscription within it and print that out as an excerpt; this
allows the effective query to be shown in the logs, which saves me from having
to look up the source file it came from.

If the regex search fails (it's not meant to be perfect), then either print out
the entire query document verbatim (if less than 500 characters), otherwise
abbreviate it to a single line as was done before.

n.b. Single-operation queries without an operation name are still printed
in full, as before.

bryan-inomial committed:
Update FINE logging example to reflect change

bryan-inomial committed:
Fix regexp in FINE logging

bryan-inomial committed:
Rework

Only TLS options feature is included in this commit.

All logging changes removed.

mlillywhite commented on InomialClient.js:
```diff
@@ -8,8 +8,8 @@
 module.exports = {
-    connect: function(hostname, stage, origin, apikey) {
-        const client = new InomialClient(hostname, stage, origin, apikey);
+    connect: function(hostname, stage, origin, apikey, websocketClientConfig) {
```
is it possible to make this a property that you set rather than a field in the constructor? the majority of users will never need this, so it should be "possible" but not get in the way of "easy".

perhaps a static function such as `InomialClient.setWebsocketClientConfig`.

bryan-inomial replied:
I've just made that change now.

bryan-inomial committed:
Move TLS options to separate static method

Removed fifth argument from InomialClient constructor and
InomialClient.connect() method, instead there is a new static method
InomialClient.setWebsocketClientConfig() which allows specifying TLS/socket
options prior to connecting.
